### PR TITLE
Formal client specification.

### PR DESCRIPTION
**Pull request by Ippo343**
This pull requests aims to implement a formal specification for the behaviour of ckan clients.
See also issue #956.

### CODE REVIEW

**Comment by Ippo343** [comment]
Current state: I wrote a starting point for v1 specification.
For the moment I am only working on version 1 of the specification: once we will have discussed and accepted version 1 (which is by far the biggest part of the work), I will work on supplementing it with subsequent extensions.

Please see the line comments for the parts where I would need some feedback.

File: client-spec/client-specification-v1.md [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+This document refers to version 1 of the CKAN metadata specification.
+
+
+
+# Definitions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in  [RFC 2119](http://www.faqs.org/rfcs/rfc2119.html).
+
+
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: TODO: Introducing the definitions for some of the terms using further down.

File: client-spec/client-specification-v1.md [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+### version
+
+A string defining the version of the module.
+The specification allows for any string to be used as a version.
+Since there is no standard regarding versioning schemes, a client should not assume any.
+
+As a significant amount of metadata in the official repository
+is obtained automatically from github, leading 'v' characters are common:
+as such, version comparison according to [PEP 440](https://www.python.org/dev/peps/pep-0440/#version-scheme) is recommended.
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: PEP 440 is nice... however, it should be noted that not even the official client implements it, so maybe recommending it is a little bold?

File: client-spec/client-specification-v1.md [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+of the same module at the same time.
+
+### download
+
+The download field contains an url where the module can be downloaded.
+This field can contain any url:
+clients must not assume this field to contain a direct link,
+but they must follow eventual redirects.
+
+The download field will always point to a zip archive:
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: ... will it?

File: client-spec/client-specification-v1.md [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+This field can have the following values:
+- `stable`
+- `testing`
+- `development`
+
+This field may be use to restrict the list of installable packages
+(e.g, to only allow `stable` modules).
+
+If this field is missing, clients must assume its value to be `stable`.
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: Is this correct?

File: client-spec/client-specification-v1.md [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+This field may contain either:
+- `any`
+- a string matching the regular expression:
+		[0-9]+\\.[0-9]+(\\.[0-9]+)?
+
+If the field contains the string `any`,
+clients must allow the installation of the module
+on any version of the game,
+unless the `ksp_version_min` and `ksp_version_max` are also present:
+in this case, the client must respect these constraints.
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: Is it possible that these fields are found together? Or should they be considered as invalid metadata?

File: client-spec/client-specification-v1.md [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+Like the `ksp_version` field, these fields can contain either the string `any`
+or a version number.
+
+If `ksp_version` is `any` and either of these fields is defined,
+the client must enforce these constraints.
+
+If `ksp_version` contains a specific version number,
+these two fields should not be present.
+In this case, the client must disregard their value and only enforce
+the most specific version constraint.
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: This looks sane to me, but I'd like to have some feedback regarding this.

File: client-spec/client-specification-v1.md [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+
+
+## install stanzas
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: What is the proper behaviour if the install stanza is missing?

File: client-spec/client-specification-v1.md [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+
+If an `install` stanza is present,
+it must contain at least two fields:
+
+- `file`: path of a file or folder, relative to the top-level of the zip file obtained through the `download` link. The file (or folder) must be extracted to the location specified by the `install_to` field.
+- `install_to`: the location where the file or folder must be extracted. This field can only take one of the following 4 values:
+	. `GameRoot`: the root folder of the KSP installation;
+	. `GameData`: equal to `$(GameRoot)/GameData`;
+    . `Ships`: equal to `$(GameRoot)/Ships`
+    . `Tutorial`: equal to #TODO: where!?
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: Where are tutorials installed, exactly? And what should be allowed there?

File: client-spec/client-specification-v1.md [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+The `filter` field only affects files in the same install stanza
+(that is, install stanzas are independent from each other).
+
+Analogously, if a `filter_regexp` is specified
+the client must exclude from the installation
+all the files that match one or more of the listed regular expressions.
+
+### File overwriting
+
+File overwriting is not permitted.
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: ... ever?

File: client-spec/client-specification-v1.md [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+Any attempt to overwrite another file
+must result in an error,
+aborting the installation attempt.
+
+### Rolling back
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: This part is a little tricky. Transactions are not a trivial thing and might not be available for every Language.
Should this constraint be forced on *every* client?

File: client-spec/client-specification-v1.md [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+
+These 4 fields are all optional.
+If present, they contain a list of relationship descriptors with the following fields:
+- `name`: contains the `identifier` of the related module (required);
+- `version`: the specific version of the related module (optional);
+- `min_version` and `max_version`: allows to specify a range of versions (optional).
+
+Version ranges are only valid for the `conflicts` relationship:
+if version ranges are encountered in other relationship descriptors,
+the behaviour is undefined: clients may choose how to resolve the ambiguity.
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: What is the proper behaviour for recommending a range of versions?

File: client-spec/client-specification-v1.md [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+this mechanism, if implemented, must be explicitly selected by the user
+for each install process.
+
+### provides
+
+Additionally, modules may indicate they provide a name.
+Providing must be used to signal interchangeable modules:
+modules may be in relationships with a provided name.
+
+Version constraints are not valid when referring to a provided name.
```
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: Is this true?

File: client-spec/client-specification-v1.md (hunk 7 above)
Comment (Dazpoet) [inline client-spec/client-specification-v1.md]: "If no install sections are provided, a CKAN client must find the top-most directory in the archive that matches the module identifier, and install that with a target of GameData." fetched from https://github.com/KSP-CKAN/CKAN/blob/master/Spec.md#install

File: client-spec/client-specification-v1.md (hunk 9 above)
Comment (Dazpoet) [inline client-spec/client-specification-v1.md]: https://github.com/KSP-CKAN/CKAN/issues/949 is probably relevant for for spec_version 1, overwriting is not allowed ever afaik. However if a client would handle updating by overwriting old files somehow... then maybe overwriting is ok? I might have missed it or does this document not handle updating of modules?

File: client-spec/client-specification-v1.md [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+
+If a package conflicts with a provided name,
+clients must consider all the packages that provide that name
+as conflicting and deal with them as described in the `conflicts` section.
```
Comment (Dazpoet) [inline client-spec/client-specification-v1.md]: I'm pretty sure the CKAN client doesn't do this? If it does why do we add both provides and conflicts to all mods who does this? Sorry for perhaps tangential question.

File: client-spec/client-specification-v1.md (hunk 3 above)
Comment (Dazpoet) [inline client-spec/client-specification-v1.md]: No, see e.g. https://github.com/KSP-CKAN/CKAN-meta/blob/master/KAS/KAS-0.4.10.ckan

File: client-spec/client-specification-v1.md (hunk 6 above)
Comment (Dazpoet) [inline client-spec/client-specification-v1.md]: Afaik the only 2 examples where both fieldtypes exists are https://github.com/KSP-CKAN/CKAN-meta/blob/master/kribubble/kribubble-1.1.ckan and https://github.com/KSP-CKAN/CKAN-meta/blob/master/krispeeder1/krispeeder1-1.0.ckan. None of them show up in my KSP 0.25 testinstall which makes it appear that `ksp_version_min` trumps `ksp_version`

Looking at the current spec.md and how we gather data from ksp-avc through [`$vref`](https://github.com/KSP-CKAN/CKAN/blob/master/Spec.md#ckanksp-avc) further enforce that `ksp_version_min` and `ksp_version_max` takes precedence over `ksp_version` in all cases.

File: client-spec/client-specification-v1.md (hunk 3 above)
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: Actually it does, if you follow the redirect. KS urls have the same behaviour: the url is that of a webpage, but it redirects you to a zip file.

Comment (Dazpoet) [inline client-spec/client-specification-v1.md]: Ah I thought you implied that the string in the field would always end in .zip. Thanks for clearing that up :)

File: client-spec/client-specification-v1.md (hunk 13 above)
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: Actually, it does: for example FAR [conflicts with AeroDynamicModel](https://github.com/KSP-CKAN/CKAN-meta/blob/master/FerramAerospaceResearch/FerramAerospaceResearch-v0.15.2_Ferri.ckan#L15-L18), which makes it incompatible with any other mod that provides that name (for example, NEAR).

Comment (Dazpoet) [inline client-spec/client-specification-v1.md]: :+1:

File: client-spec/client-specification-v1.md (hunk 2 above)
Comment (pjf) [inline client-spec/client-specification-v1.md]: There seems to be endless opinions on how we *should* handle version comparisons. The metadata spec explicitly mandates that clients *must* use the Debian version comparison rules, and the C# implementation follows these exactly.

I'm hesitant to change how we do version comparisons for two reasons:

1. We index more than 700 mods. Of those, we've only had a fraction where the Debian versioning spec has not worked, and changing the versioning implementation runs a *very real* risk of breaking existing mods.
1. The netkan pre-processor now has additional options for post-processing of versions into standard forms, or bumping the epoch when versions are truly wacky.
1. Changing versioning schemes is *much* harder than everyone seems to think. If we spot a document implementing spec 1.0–1.6, then we should use the old versioning scheme. If we spot a document with the "new" spec, we should use the new versioning scheme. If we're trying to compare *between* routines, we're screwed.

TL;DR: We're not going to change versioning schemes unless someone can provide *very* compelling reasons, a proof that we wont break existing metadata, and a sane migration plan.

TL;DR;TL;DR: We use the Debian spec.

File: client-spec/client-specification-v1.md (hunk 3 above)
Comment (pjf) [inline client-spec/client-specification-v1.md]: I wouldn't mandate that it point to a .zip file, I'd really love us to support .tar.gz and .tar.bz2 in the future, as both provide better compression and can probably even be handled with our existing libraries. They're certainly both unencumbered formats.

However clients *should* support zip files, as they're not going to be very functional without that.

File: client-spec/client-specification-v1.md (hunk 5 above)
Comment (pjf) [inline client-spec/client-specification-v1.md]: You can have:
-  `ksp_version`, OR
-  `ksp_version_min` and/or `ksp_version_max`

Under the hood, `ksp_version` is effectively a short-cut for setting both `_min` and `_max`. The form `0.23` would set `_min` to `0.23.0` and max to `0.25.MAXINT`, whereas `0.23.5` would have them both be set exactly.

File: client-spec/client-specification-v1.md (hunk 9 above)
Comment (pjf) [inline client-spec/client-specification-v1.md]: Not yet. Upgrades currently remove the old mod entirely and install the new one. Overwriting a file owned by another mod is a *hard problem*. Not only we'd have to flip the file ownership in the registry, but we'd also need to figure out what to do if that mod is uninstalled, but the mod it overwrites was not.

I need to see how Debian solves this problem; I have a hunch they require that you can only overwrite another package if you *replace it entirely*, and that this is really special magic for upgrading libc and other critical files which simply can't removed and replaced.

TL;DR: File overwriting is not permitted.

File: client-spec/client-specification-v1.md (hunk 10 above)
Comment (pjf) [inline client-spec/client-specification-v1.md]: Clients *must* leave GameData and the registry in a consistent state. How they choose to do that is up to them, but it can include checking for file conflicts before installing anything, or they can take the more lazy approach and use transactions.

So no, you don't have to have transactions, but clients are responsible for making sure things are consistent.

File: client-spec/client-specification-v1.md (hunk 11 above)
Comment (pjf) [inline client-spec/client-specification-v1.md]: I'm not sure I understand the question.

File: client-spec/client-specification-v1.md (hunk 12 above)
Comment (pjf) [inline client-spec/client-specification-v1.md]: It is! And again, this is straight out of the Debian spec. The idea is that `provides` is almost always for providing functionality¹ where there are one or more choices, but these choices may have entirely different versioning schemes underneath. I can probably dig up the relevant part of the Debian spec, as my mind seems to recall there was another monster that we avoid by doing this, but I can't remember which monster it is.

¹ Although in Debian provides is also used for package replacement, so existing relationships can be preserved when a package is obsoleted in its entirety by another package.

File: client-spec/client-specification-v1.md (hunk 13 above)
Comment (pjf) [inline client-spec/client-specification-v1.md]: Aye, Provides + Conflicts is the backbone for "you can have one, but ONLY one, of these options". It's useful for things like aerodynamic models, texture packs, career modes, and all manner of other things.

File: client-spec/client-specification-v1.md (hunk 2 above)
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: Ok, that's what we use. What should we prescribe to clients?
1. Clients MUST implement the debian version comparison, or
2. Clients SHOULD implement the debian version comparison

File: client-spec/client-specification-v1.md (hunk 3 above)
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: True that; but as of spec 1 (i.e, last november), were other archive formats allowed?
I seem to recall that originally (again, spec 1) only zip files were supported (which posed a problem with Module Manager, that had to be repackaged in a zip).
Again, IIRC, I might not.

File: client-spec/client-specification-v1.md (hunk 1 above)
Comment (Ippo343) [inline client-spec/client-specification-v1.md]: Still open after first revision.

File: client-spec/client-specification-v1.md [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+The file or folder located by the `file` field must be extracted to the destination directory
+preserving the directory tree structure in the zip file.
+Installation of any file outside of the game's directory is strictly forbidden.
+
+### Optional: filters
+
+If a `filter` field is specified,
+it contains one or more files that must be excluded from the installation
+even if they are selected by the corresponding `file` field.
+The `filter` field only affects files in the same install stanza
```
Comment (Dazpoet) [inline client-spec/client-specification-v1.md]: The `filter` field also affects folders in the same install stanza, or is that somehow included in files? Wasn't obvious to me so thought I'd point it out.

**Comment by pjf** [comment]
@Ippo343 : I'm looking to close out old PRs, and I'm not exactly sure what to do with this one. If you're still looking at building a Python core then I'm happy to review and comment, although in terms of ongoing maintenance I feel that integrating this into the existing metadata spec would ensure it continues to be updated going forward.

Alternatively, if you're happy to close this out, then that's one step  closer to my goal of PR-box zero. :)

File: client-spec/client-specification-v1.md [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+(that is, install stanzas are independent from each other).
+
+Analogously, if a `filter_regexp` is specified
+the client must exclude from the installation
+all the files that match one or more of the listed regular expressions.
+
+### File overwriting
```
Comment (plague006) [inline client-spec/client-specification-v1.md]: We might want to remove this entry. Issues like https://github.com/KSP-CKAN/CKAN/issues/1311 show that we have thought about (and generally support) a limited method for overwriting so I don't see why we would impose this restriction on others.

**Comment by Ippo343** [comment]
Hey Paul, I've moved on to other things, so (sadly) I don't think I will continue with the python client. This can be closed safely.
________________________________
Da: Paul Fenwick<mailto:notifications@github.com>
Inviato: ‎13/‎12/‎2015 07:18
A: KSP-CKAN/CKAN<mailto:CKAN@noreply.github.com>
Cc: Michele<mailto:m.ippolito@outlook.com>
Oggetto: Re: [CKAN] Formal client specification. (#961)

@Ippo343 : I'm looking to close out old PRs, and I'm not exactly sure what to do with this one. If you're still looking at building a Python core then I'm happy to review and comment, although in terms of ongoing maintenance I feel that integrating this into the existing metadata spec would ensure it continues to be updated going forward.

Alternatively, if you're happy to close this out, then that's one step  closer to my goal of PR-box zero. :)

---

**Comment by pjf** [comment]
> Hey Paul, I've moved on to other things, so (sadly) I don't think I will continue with the python client. 

This is code for "being an awesome guitarist", yes? :)

> This can be closed safely. 

Roger that! One step closer to PR-box zero!

P.S. You're amazing!

**Comment by Ippo343** [comment]
This is code for "being an awesome guitarist", yes? :)

AND skydiver AND cook :P And watch out next year for the paragliding course ;)

    P.S. You're amazing!

Nah, I just have a good role model to follow, that being you :)
